pkg/query/build_tree_test.go

```diff
@@ -0,0 +1,363 @@
+			},
+		},
+	}
+
+	objectDefs := []*corev1.NamespaceDefinition{userDef, groupDef}
+	dsSchema, err := schema.BuildSchemaFromDefinitions(objectDefs, nil)
+	require.NoError(err)
+
+	// This should detect recursion and return an error
+	_, err = query.BuildIteratorFromSchema(dsSchema, "group", "member")
+	require.Error(err)
+	require.Contains(err.Error(), "recursive schema iterators are as yet unsupported")
+}
+
+func TestBuildTreeArrowOperation(t *testing.T) {
+	t.Parallel()
+
+	require := require.New(t)
+
+	// Test that we can detect when arrow operations would be created
+	// by examining a complex permission that would include them
+	objectDefs := []*corev1.NamespaceDefinition{testfixtures.UserNS.CloneVT(), testfixtures.FolderNS.CloneVT(), testfixtures.DocumentNS.CloneVT()}
+	dsSchema, err := schema.BuildSchemaFromDefinitions(objectDefs, nil)
+	require.NoError(err)
+
+	// The schema conversion should successfully build the schema
+	// even if we can't directly test arrow execution due to recursion
+	require.NotNil(dsSchema)
+
+	// Verify that the schema contains definitions with arrow-like operations
+	docDef, ok := dsSchema.Definitions["document"]
+	require.True(ok, "should find document definition")
+
+	viewPerm, ok := docDef.Permissions["view"]
+	require.True(ok, "should find view permission that contains arrow operations")
+	require.NotNil(viewPerm.Operation, "view permission should have operation defined")
+}
+
+func TestBuildTreeIntersectionOperation(t *testing.T) {
+	t.Parallel()
+
+	require := require.New(t)
+	rawDS, err := dsfortesting.NewMemDBDatastoreForTesting(0, 0, memdb.DisableGC)
+	require.NoError(err)
+
+	ds, revision := testfixtures.StandardDatastoreWithData(rawDS, require)
+
+	objectDefs := []*corev1.NamespaceDefinition{testfixtures.UserNS.CloneVT(), testfixtures.FolderNS.CloneVT(), testfixtures.DocumentNS.CloneVT()}
+	dsSchema, err := schema.BuildSchemaFromDefinitions(objectDefs, nil)
+	require.NoError(err)
+
+	// Test building iterator for view_and_edit permission which uses intersection operations
+	it, err := query.BuildIteratorFromSchema(dsSchema, "document", "view_and_edit")
+	require.NoError(err)
+
+	// Should create an intersection iterator
+	explain := it.Explain()
+	require.NotEmpty(explain.String())
+	require.Contains(explain.String(), "Intersection", "should create intersection iterator")
+
+	ctx := &query.Context{
+		Context:   t.Context(),
+		Datastore: ds,
+		Revision:  revision,
+	}
+
+	// Test execution
+	relSeq, err := it.Check(ctx, []string{"specialplan"}, "multiroleguy")
+	require.NoError(err)
+
+	rels, err := query.CollectAll(relSeq)
+	require.NoError(err)
+	t.Logf("Intersection operation results: %+v", rels)
+}
+
+func TestBuildTreeExclusionOperation(t *testing.T) {
+	t.Parallel()
+
+	require := require.New(t)
+
+	// Create a simple schema to test exclusion handling
+	userDef := testfixtures.UserNS.CloneVT()
+
+	// Create a document with an exclusion operation
+	docDef := &corev1.NamespaceDefinition{
```

Comment:
I'd recommend using the namespace builders here and in the rest of the tests: https://github.com/authzed/spicedb/blob/ebd258d28cb40648490227d6c32ffcf74295bd2f/pkg/namespace/builder.go#L4